### PR TITLE
Implement PoA Update Generator Key

### LINKED ISSUE #8416: Implement Update Generator Key command for PoA module
### Description

Implement update generator key command as described in the [LIP](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0047.md#update-generator-key-command)

### Acceptance Criteria
- Add unit tests covering all the path.

### PR DESCRIPTION
**Pull request by Phanco**
### What was the problem?

This PR resolves #8416

### How was it solved?
Implemented `update_generator_key.ts`

### How was it tested?
Unit Test Updated

### CODE REVIEW

**Review by sergeyshemyakov — APPROVED** [review]
I have reviewed only Update Generator Key Command, the rest is reviewed in [another PR](https://github.com/LiskHQ/lisk-sdk/pull/8527).

File: framework/src/modules/poa/commands/update_authority.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+					`${address.getLisk32AddressFromPublicKey(
+						newValidator.address,
+					)} is not a valid validator.`,
+				);
+			}
+			totalWeight += newValidator.weight;
+		}
+
+		if (totalWeight > MAX_UINT64) {
+			throw new Error('TotalWeight out of range.');
```
Comment (ishantiw) [inline framework/src/modules/poa/commands/update_authority.ts]: maybe we can give more informative error messages
```suggestion
			throw new Error('TotalWeight should be less than MAX_UINT64.');
```

File: framework/src/modules/poa/commands/update_authority.ts [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+			aggregationBits,
+			signature,
+			MESSAGE_TAG_POA,
+			context.chainID,
+			message,
+			validatorInfos.map(validatorInfo => validatorInfo.weight),
+			snapshotStore.threshold,
+		);
+		if (!verified) {
+			this.events.get(AuthorityUpdateEvent).log(context, {
```
Comment (ishantiw) [inline framework/src/modules/poa/commands/update_authority.ts]: I think you need to emit a persistent event here

File: framework/src/modules/poa/commands/update_authority.ts [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+				result: UpdateAuthority.FAIL_INVALID_SIGNATURE,
+			});
+			throw new Error('Invalid Signature.');
```
Comment (ishantiw) [inline framework/src/modules/poa/commands/update_authority.ts:154]: maybe elaborate a bit more on this error, like `Invalid weighted aggregated signature`

File: framework/src/modules/poa/commands/update_authority.ts [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+	): Promise<void> {
+		const { newValidators, threshold, validatorsUpdateNonce, aggregationBits, signature } =
+			context.params;
+		const message = codec.encode(validatorSignatureMessageSchema, {
+			newValidators,
+			threshold,
+			validatorsUpdateNonce,
+		});
+		const validatorInfos = [];
+		const snapshotStore = await this.stores.get(SnapshotStore).get(context, Buffer.from([0]));
```
Comment (ishantiw) [inline framework/src/modules/poa/commands/update_authority.ts]: I think you need to use `intToBuffer` method from cryptography.utils here for the 4 byte key

```suggestion
		const snapshotStore = await this.stores.get(SnapshotStore).get(context, intToBuffer(0, 4));
```

File: framework/src/modules/poa/commands/update_authority.ts [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+			validatorInfos.map(validatorInfo => validatorInfo.weight),
+			snapshotStore.threshold,
+		);
+		if (!verified) {
+			this.events.get(AuthorityUpdateEvent).log(context, {
+				result: UpdateAuthority.FAIL_INVALID_SIGNATURE,
+			});
+			throw new Error('Invalid Signature.');
+		}
+		await this.stores.get(SnapshotStore).set(context, Buffer.from([2]), {
```
Comment (ishantiw) [inline framework/src/modules/poa/commands/update_authority.ts]: same here , use `intToBuffer(2, 4)`

**Review by ishantiw — CHANGES_REQUESTED** [review]

File: framework/src/modules/poa/commands/update_generator_key.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+		} catch (err) {
+			return {
+				status: VerifyStatus.FAIL,
+				error: err as Error,
+			};
+		}
+
+		const validatorExists = await this.stores
+			.get(ValidatorStore)
+			.has(context, context.transaction.senderAddress);
```
[the remark this answers is not in the document]
Comment (Phanco) [reply framework/src/modules/poa/commands/update_generator_key.ts:55]: There're the same, now both using class function `context.transaction.senderAddress`

File: framework/src/modules/poa/commands/update_authority.ts (hunk 1 above)
Comment (Phanco) [reply framework/src/modules/poa/commands/update_authority.ts]: Updated on https://github.com/LiskHQ/lisk-sdk/pull/8527

File: framework/src/modules/poa/commands/update_authority.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+			status: VerifyStatus.OK,
+		};
+	}
+
+	public async execute(
+		context: CommandExecuteContext<UpdateAuthorityValidatorParams>,
+	): Promise<void> {
+		const { newValidators, threshold, validatorsUpdateNonce, aggregationBits, signature } =
+			context.params;
+		const message = codec.encode(validatorSignatureMessageSchema, {
```
[the remark this answers is not in the document]
Comment (Phanco) [reply framework/src/modules/poa/commands/update_authority.ts:113]: Updated on https://github.com/LiskHQ/lisk-sdk/pull/8527

**Review by ishantiw — APPROVED** [review]

**Pull request merged**